### PR TITLE
1135/DeFi-R. User Flow Optimization: Two Flows Support

### PR DESCRIPTION
**Pull request by HaidarJbeily7**
## Changes

- Added regulate_asset extrinsic with benchmarking and unit tests

### CODE REVIEW

File: pallets/extended-assets/src/lib.rs [hunk 1]
```diff
@@ -341,6 +341,35 @@ pub mod pallet {
...
+        /// Marks an asset as regulated, representing that the asset will only operate between KYC-verified wallets.
+        ///
+        /// ## Parameters
+        ///
+        /// - `origin`: The origin of the transaction.
+        /// - `asset_id`: The identifier of the asset.
+        #[pallet::call_index(4)]
+        #[pallet::weight(<T as Config>::WeightInfo::regulate_asset())]
+        pub fn regulate_asset(origin: OriginFor<T>, asset_id: AssetIdOf<T>) -> DispatchResult {
+            let who = ensure_signed(origin)?;
```
Comment (wer1st) [inline pallets/extended-assets/src/lib.rs:354]: Do we want to allow this for `sudo` as well?

File: pallets/extended-assets/src/lib.rs [hunk 2]
```diff
@@ -341,6 +341,35 @@ pub mod pallet {
...
+
+        /// Marks an asset as regulated, representing that the asset will only operate between KYC-verified wallets.
+        ///
+        /// ## Parameters
+        ///
+        /// - `origin`: The origin of the transaction.
+        /// - `asset_id`: The identifier of the asset.
+        #[pallet::call_index(4)]
+        #[pallet::weight(<T as Config>::WeightInfo::regulate_asset())]
+        pub fn regulate_asset(origin: OriginFor<T>, asset_id: AssetIdOf<T>) -> DispatchResult {
```
Comment (wer1st) [inline pallets/extended-assets/src/lib.rs:353]: As I remember, the main idea is to allow to operate with the asset only for KYC passed users.
What if the asset is registered and distributed between non-KYC users and then after that we will regulate the asset?
In this case we have non-KYC users with regulated asset.

File: pallets/extended-assets/src/lib.rs (hunk 1 above)
Comment (HaidarJbeily7) [reply pallets/extended-assets/src/lib.rs:354]: There is no such requirement so, I don't think so. it's just in the scope of asset owner.

File: pallets/extended-assets/src/lib.rs (hunk 2 above)
Comment (HaidarJbeily7) [reply pallets/extended-assets/src/lib.rs:353]: Yes, it's possible for asset owner to regulate the asset after distribution.
Simply enough they can force the users to verify KYC by regulating the asset in order to operate further.

**Review by vovac12 — APPROVED** [review]

File: pallets/extended-assets/src/lib.rs (hunk 1 above)
Comment (wer1st) [reply pallets/extended-assets/src/lib.rs:354]: We have to address the question to requirements.

Comment (HaidarJbeily7) [reply pallets/extended-assets/src/lib.rs:354]: @AndreyKortosov 
Can you please address this in requirements?

The question:
Do we need the sudo (root) account to have access to regulate any asset in addition to AssetOwner account?


cc. @wer1st

Comment (AndreyKortosov) [reply pallets/extended-assets/src/lib.rs:354]: @HaidarJbeily7 AssetOwner account will regulate any asset. Am I right?
What's the purpose of sharing access to another account?

Comment (HaidarJbeily7) [reply pallets/extended-assets/src/lib.rs:354]: @AndreyKortosov 
AssetOwner account will be able to regulate **the assets that the account owns**.
There is a sudo account which has superior accesses. So, we need to address in requirements if we need to share access with this sudo account to this extrinsic, or simply not?

@vovac12 Do you think there is a use case that we would need sudo access?

Comment (wer1st) [reply pallets/extended-assets/src/lib.rs:354]: Ok

**Review by wer1st — APPROVED** [review]

**Pull request merged**